sdk/azidentity/chained_token_credential.go

```diff
@@ -50,35 +53,73 @@ func NewChainedTokenCredential(sources []azcore.TokenCredential, options *Chaine
 	if options == nil {
 		options = &ChainedTokenCredentialOptions{}
 	}
-	return &ChainedTokenCredential{sources: cp, name: "ChainedTokenCredential", retrySources: options.RetrySources}, nil
+	return &ChainedTokenCredential{
+		cond:         sync.NewCond(&sync.Mutex{}),
+		name:         "ChainedTokenCredential",
+		retrySources: options.RetrySources,
+		sources:      cp,
+	}, nil
 }
 
 // GetToken calls GetToken on the chained credentials in turn, stopping when one returns a token. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
-	if c.successfulCredential != nil && !c.retrySources {
-		return c.successfulCredential.GetToken(ctx, opts)
+	if !c.retrySources {
+		// ensure only one goroutine at a time iterates the sources and perhaps sets c.successfulCredential
+		c.cond.L.Lock()
+		for {
+			if c.successfulCredential != nil {
+				c.cond.L.Unlock()
+				return c.successfulCredential.GetToken(ctx, opts)
+			}
+			if !c.iterating {
+				c.iterating = true
+				// allow other goroutines to wait while this one iterates
+				c.cond.L.Unlock()
+				break
+			}
+			c.cond.Wait()
+		}
 	}
 
+	var err error
 	var errs []error
+	var token *azcore.AccessToken
 	for _, cred := range c.sources {
-		token, err := cred.GetToken(ctx, opts)
+		token, err = cred.GetToken(ctx, opts)
 		if err == nil {
 			log.Writef(EventAuthentication, "%s authenticated with %s", c.name, extractCredentialName(cred))
-			c.successfulCredential = cred
-			return token, nil
+			if c.iterating {
+				c.cond.L.Lock()
+				c.successfulCredential = cred
+				c.cond.L.Unlock()
```

Comment:
I think this unlock could be removed, and the lock on line 106.  What do you think?

Reply:
Works for the happy path but a goroutine that exits this loop due to an error doesn't execute this block. Such a goroutine would then go on to assign `c.iterating` and unlock while not holding the lock. This would be easier to reason about and a bit tidier without this block, let's try that.